File: matlab/encode_ply.m
%run RAHT + RLGR coder for attributes
clear;
addpath(genpath('RA-GFT'));
addpath(genpath('RAHT'));

function J_adj = raht_adjust_depth_increase_only(V, minV, width, J_req)
% Choose a depth that fits data; only increases J if needed.
% V: N×3 double, minV: 1×3, width: scalar (cube side), J_req: requested depth

    % voxel size for requested depth
    Q = width / 2^J_req;

    % quantize with a tiny tolerance to avoid boundary spillover
    Vint = floor( (V - minV) / Q - 1e-12 );

    % maximum index across all axes
    maxIndex = max(Vint(:));              % could be negative if minV>min(V), but that’s fine
    needed   = ceil(log2(double(maxIndex + 1)));  % bits needed per axis

    % if data has no extent (needed<=0), depth=0 is enough
    needed = max(0, needed);

    % only increase if needed
    J_adj = max(J_req, needed);

    % cap for uint64 Morton codes (3*J <= 63)
    J_adj = min(J_adj, 21);
end

% List of PLY file paths
ply_paths = {'/ssd1/haodongw/workspace/3dstream/3DGS_Compression_Adaptive_Voxelization/attributes_compressed/train_depth_15_thr_30_3DGS_adapt_lossless/train_dc.ply'};
J = 15;

T = length(ply_paths);
Nvox = zeros(T,1);
time = Nvox;

for frame =1:T
    tic;

    [ V, C ] = read_ply_file(ply_paths{frame});
    N = size(V,1);
    Nvox(frame) = N;

    J_adj = raht_adjust_depth_increase_only(V, [0 0 0], 2^J, J);

    % ListC: indices of colors
    % FlagsC: indicator of whether a node is a left sibling to another node
    % weightsC: weights of colors
    [ListC,FlagsC,weightsC]=RAHT_param(V,[0 0 0], 2^J_adj, J_adj);
    
    [Coeff,w]=RAHT(C,ListC,FlagsC,weightsC);
    C_recon = iRAHT(Coeff, ListC, FlagsC, weightsC);

    % Save for debug
    save(sprintf('../results/frame%d_coeff_matlab.mat', frame), 'Coeff');
    save(sprintf('../results/frame%d_params_matlab.mat', frame), 'ListC', 'FlagsC', 'weightsC');

    if ismembertol(C, C_recon, 1e-8)
        fprintf('Reconstruction check: PASSED (within tolerance)\n');
    else
        fprintf('Reconstruction check: FAILED (difference exceeds tolerance)\n');
    end

    fprintf('L2-norm of C: %.4f\n', norm(C));
    fprintf('L2-norm of Coeff: %.4f\n', norm(Coeff));
    time(frame)=toc;
    disp([time(frame) frame T])
end

File: python/encode_ply.py
import torch
import numpy as np
import time
import matplotlib.pyplot as plt
from scipy.io import loadmat

from data_util import read_ply_file
from utils import save_mat, save_lists
from RAHT import RAHT2, RAHT2_optimized, RAHT_batched
from iRAHT import inverse_RAHT
from RAHT_param import RAHT_param
import rlgr

VARIANTS = {
    "RAHT":             lambda C,L,F,W,d: RAHT2(C, L, F, W, d),
    "RAHT_optimized":   lambda C,L,F,W,d: RAHT2_optimized(C, L, F, W, d),
    "RAHT_batched":     lambda C,L,F,W,d: RAHT_batched(C, L, F, W, d),
}

DEBUG = True

def sanity_check_vector(T: torch.Tensor, C: torch.Tensor, rtol=1e-5, atol=1e-8) -> bool:
    """
    Sanity check: max(T) == sqrt(N) * mean(C)
    T, C: 1D tensors of shape [N]
    """
    assert T.dim() == 1 and C.dim() == 1 and T.size(0) == C.size(0), "T and C must be 1D with same length"
    N = T.size(0)

    lhs = T.max()
    rhs = torch.sqrt(torch.tensor(float(N), dtype=C.dtype, device=C.device)) * C.mean()

    return torch.allclose(lhs, rhs, rtol=rtol, atol=atol)


## ---------------------
## Configuration
## ---------------------
ply_list = ['/ssd1/haodongw/workspace/3dstream/3DGS_Compression_Adaptive_Voxelization/attributes_compressed/train_depth_15_thr_30_3DGS_adapt_lossless/train_dc.ply']
J = 18
T = len(ply_list)

device = 'cuda:1' if torch.cuda.is_available() else 'cpu'

colorStep = [1, 2, 4, 6, 8, 12, 16, 20, 24, 32, 64]
nSteps = len(colorStep)
bytes_log = torch.zeros((T, nSteps))
MSE = torch.zeros((T, nSteps))
Nvox = torch.zeros(T)
time_log = torch.zeros(T)

## ---------------------
## Main Processing Loop
## ---------------------
for frame_idx in range(T):
    frame = frame_idx + 1
    frame_start = time.time()
    
    V, Crgb = read_ply_file(ply_list[frame_idx])
    N = V.shape[0]
    Nvox[frame_idx] = N
    C = Crgb
    
    origin = torch.tensor([0, 0, 0], dtype=V.dtype)
    t0 = time.time()
    ListC, FlagsC, weightsC = RAHT_param(V, origin, 2**J, J)
    t1 = time.time()
    raht_param_time = t1 - t0
    
    ListC = [t.to(device) for t in ListC]
    FlagsC = [t.to(device) for t in FlagsC]
    weightsC = [t.to(device) for t in weightsC]
    C = C.to(device)
    
    timings = {}
    coeffs_by_variant = {}
    for name, fn in VARIANTS.items():
        if torch.cuda.is_available():
            torch.cuda.synchronize()
        t2 = time.time()
        Coeff, w = fn(C, ListC, FlagsC, weightsC, device)
        if torch.cuda.is_available():
            torch.cuda.synchronize()
        t3 = time.time()
        timings[name] = t3 - t2
        coeffs_by_variant[name] = Coeff
        
        if DEBUG:
            print(f"Norm of C: {torch.norm(C)}")
            print(f"Norm of Coeff: {torch.norm(Coeff)}")
            print(f"Sanity check: {sanity_check_vector(Coeff[:, 0], C[:, 0])}")
            print(f"Sanity check: {sanity_check_vector(Coeff[:, 1], C[:, 1])}")
            print(f"Sanity check: {sanity_check_vector(Coeff[:, 2], C[:, 2])}")
            C_recon = inverse_RAHT(Coeff, ListC, FlagsC, weightsC, device)
            if torch.allclose(C, C_recon, rtol=1e-4, atol=1e-6):
                print("Reconstruction check: True")
            else:
                print("Reconstruction check: False")
                diff = C - C_recon
                print("Max difference:", diff.abs().max().item())
                print("Mean difference:", diff.abs().mean().item())
        
        # Sort weights in descending order
        _, IX_ref = torch.sort(w, descending=True)
        Y = Coeff[:, 0]
        
        # temporary: filename for PyRLGR
        filename = 'test.bin'
        
        for i in range(nSteps):
            step = colorStep[i]
            Coeff_enc = torch.round(Coeff / step)
            Y_hat = Coeff_enc[:, 0] * step
            
            MSE[frame_idx, i] = (torch.linalg.norm(Y - Y_hat)**2) / (N * 255**2)
            
            # nbytesY, _ = RLGR_encoder(Coeff_enc[IX_ref, 0])
            # nbytesU, _ = RLGR_encoder(Coeff_enc[IX_ref, 1])
            # nbytesV, _ = RLGR_encoder(Coeff_enc[IX_ref, 2])
            # bytes_log[frame_idx, i] = nbytesY + nbytesU + nbytesV
            
            enc = rlgr.file(filename, 1)
            Y_list = [int(i) for i in Coeff_enc[IX_ref, 0].squeeze(1).tolist()]
            U_list = [int(i) for i in Coeff_enc[IX_ref, 1].squeeze(1).tolist()]
            V_list = [int(i) for i in Coeff_enc[IX_ref, 2].squeeze(1).tolist()]
            enc.rlgrWrite(Y_list, 0)
            enc.rlgrWrite(U_list, 0)
            enc.rlgrWrite(V_list, 0)
            enc.close()

    # Print timing info
    print(f"Frame {frame}: RAHT_param={raht_param_time:.6f}s")
    for name, t in timings.items():
        print(f"  {name}: {t:.6f}s")
    
    time_log[frame_idx] = time.time() - frame_start
    print(f"  Frame {frame}/{T} processed in {time_log[frame_idx]:.2f} seconds.")

# ## ---------------------
# ## Analysis, Plotting, and Saving
# ## ---------------------
# print("Analyzing results...")

# # Calculate PSNR from the mean MSE across all frames for each quantization step
# psnr = -10 * torch.log10(torch.mean(MSE, dim=0))

# # Calculate bits per voxel (bpv)
# total_bytes_per_step = torch.sum(bytes_log, dim=0)
# total_voxels = torch.sum(Nvox)
# bpv = 8 * total_bytes_per_step / total_voxels

# # --- Plotting ---
# plt.figure(figsize=(8, 6))
# plt.plot(bpv.numpy(), psnr.numpy(), 'b-x', label='O3D Load + RAHT Sim')
# plt.xlabel('Bits per Voxel (bpv)')
# plt.ylabel('Y-PSNR (dB)')
# plt.title('Rate-Distortion Curve')
# plt.grid(True)
# plt.legend()
# plt.show()
